config/config.go

```
@@ -106,69 +106,86 @@ func parsePlugins(pluginList []interface{}) ([]*PluginConfig, error) {
+	if i := strings.LastIndexByte(ip, '%'); i >= 0 {
+		ip, zone = ip[:i], ip[i+1:]
 	}
-	return ifname, nil
+	return
 }
 
-func (c *Config) getListenAddress(ver protocolVersion) (*net.UDPAddr, error) {
+func (c *Config) getListenAddress(addr string, ver protocolVersion) (*net.UDPAddr, error) {
 	if err := protoVersionCheck(ver); err != nil {
 		return nil, err
 	}
-	addr := c.v.GetString(fmt.Sprintf("server%d.listen", ver))
-	if addr == "" {
-		// return default listener
-		if ver == protocolV6 {
-			return &net.UDPAddr{
-				IP:   net.IPv6unspecified,
-				Port: dhcpv6.DefaultServerPort,
-			}, nil
-		}
-		return &net.UDPAddr{
-			IP:   net.IPv4zero,
-			Port: dhcpv4.ServerPort,
-		}, nil
-	}
-	ipStr, portStr, err := net.SplitHostPort(addr)
+
+	ipStr, ifname, portStr, err := splitHostPort(addr)
 	if err != nil {
 		return nil, ConfigErrorFromString("dhcpv%d: %v", ver, err)
 	}
+
 	ip := net.ParseIP(ipStr)
+	if ipStr == "" {
+		switch ver {
+		case protocolV4:
+			ip = net.IPv4zero
+		case protocolV6:
+			ip = net.IPv6unspecified
+		}
+	}
 	if ip == nil {
-		return nil, ConfigErrorFromString("dhcpv%d: invalid IP address in `listen` directive", ver)
+		return nil, ConfigErrorFromString("dhcpv%d: invalid IP address in `listen` directive: %s", ver, ipStr)
 	}
-	if ver == protocolV6 && ip.To4() != nil {
-		return nil, ConfigErrorFromString("dhcpv%d: not a valid IPv6 address in `listen` directive", ver)
-	} else if ver == protocolV4 && ip.To4() == nil {
-		return nil, ConfigErrorFromString("dhcpv%d: not a valid IPv4 address in `listen` directive", ver)
+	if ip4 := ip.To4(); (ver == protocolV6 && ip4 != nil) || (ver == protocolV4 && ip4 == nil) {
+		return nil, ConfigErrorFromString("dhcpv%d: not a valid IPv%d address in `listen` directive", ver, ver)
 	}
-	port, err := strconv.Atoi(portStr)
-	if err != nil {
-		return nil, ConfigErrorFromString("dhcpv%d: invalid `listen` port", ver)
+
+	var port int
+	if portStr == "" {
+		switch ver {
```

Same here about default

---

Same as above, taking the default would be a bug (`ver` doesn't come from user input), we can signal explicitly on that. 
This one is a bit different as it would result in port 0 being silently used instead of an error which is a bigger issue; I can add it

---

yeah, and as above I'd `panic()` here, to be loud about the bug